### PR TITLE
Ci/add jobs to travis continous integration configuration

### PR DESCRIPTION
**Pull request by patrikgrenfeldt**
This PR adds jobs to the travis builds.  
 🤕 The job checking that the code quality score was not decreased relies on a .configs/pylintrc on master a wont succeed until merged onto master. I suggest we think that is OK?

**How to test push jobs**:
- [x] open the travis push check 

**Expected outcome**:
- [x] There should be a coverage job

**How to test PR jobs**:
- [x] open the travis PR check 

**Expected outcome**:
- [x] There should be a job running tests with python 3.6
- [x] There should be a job running tests with python 3.7
- [x] There should be a job checking that the package is installable via setup.py
- [x] There should be a job checking installed prod deps w python 3.6
- [x] There should be a job checking installed prod deps w python 3.7
- [x] There should be a job checking installed dev deps
- [x] There should be a job checking the code coverage
- [x] There should be a job checking the code formatting
- [x] There should be a job checking that the code quality score was not decreased. 🤕  This job relies on a .configs/pylintrc on master a wont succeed until merged onto master 
- [x] There should be a job linting other files than .py

Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by@dnil @adrosenbaum 
- [x] tests executed by @patrikgrenfeldt

### CODE REVIEW

**Comment by patrikgrenfeldt** [comment]
Test results pust:
image
👍 
Test result PR checks:
image
👍

**Comment by northwestwitch** [comment]
Yes, thanks for the cool improvements @patrikgrenfeldt! After merging this I have a suggestion for the dumbest PR ever ---> black everything so we'll have a baseline of good code from which develop every new PR. This way it will be much easier to review the code, because we won't see all those changes introduces by the formatting itself..

**Comment by ingkebil** [comment]
@dnil regarding the license: @vwirta has access to legal to look into these matters if needed.

**Comment by northwestwitch** [comment]
> @northwestwitch I really like the idea maybe you could make a pull request with an all blacked Scout?

yess!

image

**Comment by dnil** [comment]
😆 Nice! I will readily admit that not conforming to PEP8, pycodestyle or what not is not keeping me sleepless, but small diffs are good, and consistent style is sort of nice. Unless it involves missing whitespaces, trailing commas or unaligned parenthesis, naturally.

**Comment by dnil** [comment]
A quick read-through of the git-lint docs appears to indicate a default config is already provided with the installation of the package. As there are no change notices on the file included, I'm guessing it is redundant? Or are the docs just outdated?

**Comment by patrikgrenfeldt** [comment]
We have changed the allowed line lengths in the config as well as how pylint behaves (only report errors). We have also removed most configurations for filetypes we don't expect to encounter.

**Comment by dnil** [comment]
- [ ] still has ```# Copyright 2013-2014 Sebastian Kreft```, in a file under Apache license that has been modified, but where there is no notice of this in the file. What was the issue with using the package default, except that it then lints jpeg and what-not?
- [ ] use a couple of sensible defaults for the travis builds. We can't keep updating these files in every repo every time an env changes on hasta or clinical-db - or their heirs. That belongs in a system test repo (eg servers for us at CG Solna) and not in the component repos.

**Comment by patrikgrenfeldt** [comment]
I have two options I'd like to suggest:
We include a copy of the original config file with a notice that all changes can be found via a link to its own file history on GitHub or
2. We bake our own config
Thoughts?

**Comment by patrikgrenfeldt** [comment]
>     * [ ]  still has `# Copyright 2013-2014 Sebastian Kreft`, in a file under Apache license that has been modified, but where there is no notice of this in the file. If there is no way to add a small config with just our local edits, then we should add a change note to the file to comply with its license.
> 
>     * [ ]  use a couple of sensible, generally named defaults for the travis builds. We can't keep updating these files in every repo every time a conda env changes on hasta or clinical-db - or their heirs. That belongs in a system test repo (eg servers for us at CG Solna) and not in the component repos.

Just read your comment again and that was exactly your thoughts it seems 😊

**Comment by dnil** [comment]
Sounds fine! Ill drop by in a bit as well so we can talk about it!

**Comment by northwestwitch** [comment]
Ill be at the office around 9:35, can I join the discussion?

**Comment by northwestwitch** [comment]
>  pretty sure we can live without syntax checking of JPEGs and PNGs. Perhaps we could just skip that part for now, or make do with a simple lint for languages that we write ourselves?

I'd also avoid linting of HTML files, for readability issues. Maybe after the linting they would be more readable but since I am not sure I would't risk flatten up their structure after we reached such a nice consent over the years.. What do you think?

**Comment by northwestwitch** [comment]
I agree with @dnil that every reference and setting specific to our own servers should remain out of the config files. This software is supposed to be used also by other institutes, and most importantly we want other external users to contribute to the code, so I'd also avoid making the checks very strict, since this would make external pull requests very difficult to make.

**Comment by dnil** [comment]
Wow, that's quick! Thank you @patrikgrenfeldt! 🥇 Let me know when to have a look again.
@northwestwitch and @moonso got a hold of me and have educated me a bit more, so I am now really looking forward to using it!

Four things came up:
- we should add a notice to the contributor instructions briefly describing the tests, and stating that while we strive to keep things as kosher as possible, we value their contributions and will look at and merge code even if it is not 100% PEP8 compliant, blacked and whatnot. Scaring an already small community away would be worse than a few improperly formatted lines of code.
- Is there a way to have build badges state that tests passed and formatting failed separately, since we will likely have stylewise flawed builds for a while longer?
- what is your opinion, should we black the python code before or after merging this PR? Any difference?
- oh, and the new pylintrc, its not just the same without the copyright notice, right? 😬

**Comment by patrikgrenfeldt** [comment]
I don't think this will affect the badges at all (asuming the badges work towards master and not PR's). 

I would do the total black scout in a separate PR to keep separate concerns separately. It might also be fruitful to not black all of scout at once to avoid the discussion of if all of scout needs retesting after all files were changed. Black formatting all of scout at once would probably make the contributions go a bit smoother since the code formatting should have less suggestions about code formatting.

**Comment by patrikgrenfeldt** [comment]
I think it is done now. Could you have a look again @dnil ?

**Comment by moonso** [comment]
Hi @patrikgrenfeldt nice job!! I think we should black all of scout in one go asap. Is anyone up for doing that? I know you did it once for CG ;)

**Comment by northwestwitch** [comment]
> Hi @patrikgrenfeldt nice job!! I think we should black all of scout in one go asap. Is anyone up for doing that? I know you did it once for CG ;)

I agree, let's black it!

**Comment by northwestwitch** [comment]
I think we should black Scout first, then merge this PR, so it won't be complaining too much after, what do you think?

**Comment by dnil** [comment]
All for blacking as long as we agree that it in theory and established practice returns functionally equivalent code. If both theory, previous experience, automated tests and a regular stage run-through give no hints function has changed, I can't see a need for revalidation. We wouldn't categorise that as a major version change.

Before blacking we should first merge all pending PRs that are at least in some kind of shape. There would be epic amounts of conflicts otherwise, right?

File: .travis.yml [hunk 1]
```diff
@@ -1,24 +1,70 @@
----
 language: python
 
-python:
-    - "3.6"
+cache: pip
+env: TRAVIS=1
```
Comment (dnil) [inline .travis.yml]: I did like the different versions, but @northwestwitch is probably right. Let's keep it simple. There will be an option to change this when switching to git tasks at some point.

**Comment by dnil** [comment]
Oh, and before merging would you 
- add that small text to the contributions doc
- update CHANGELOG.

**Comment by northwestwitch** [comment]
> All for blacking as long as we agree that it in theory and established practice returns functionally equivalent code. If both theory, previous experience, automated tests and a regular stage run-through give no hints function has changed, I can't see a need for revalidation. We wouldn't categorise that as a major version change.
> 
> Before blacking we should first merge all pending PRs that are at least in some kind of shape. There would be epic amounts of conflicts otherwise, right?

Agreed!

File: .travis.yml (hunk 1 above)
Comment (patrikgrenfeldt) [reply .travis.yml]: Then maybe you would have liked [this](https://travis-ci.org/Clinical-Genomics/scout/builds/637346864)?

File: .travis.yml [hunk 2]
```diff
@@ -1,24 +1,70 @@
...
+        - name: "Pylint score"
+          if: type = pull_request
+          install: pip install pylint
+          script:
+              - grep -r -E "pylint. {0,1}disable\=" .; if [ $? -eq 0 ]; then echo "Can not run pylint scoring with any pylint warnings disabled, please remove them and try again" && false; else true; fi
+              - (git --no-pager diff --name-only --diff-filter=M $TRAVIS_COMMIT_RANGE | grep -F ".py" || echo "$(basename "$PWD")") > $TRAVIS_HOME/before_files.txt
+              - (git --no-pager diff --name-only --diff-filter=AM $TRAVIS_COMMIT_RANGE | grep -F ".py" || echo "$(basename "$PWD")") > $TRAVIS_HOME/after_files.txt
+              - pylint --rcfile=.configs/pylintrc --jobs=0 --exit-zero $(< $TRAVIS_HOME/after_files.txt) > $TRAVIS_HOME/pylint_after_output.txt
+              - git checkout $TRAVIS_BRANCH
+              - pylint --rcfile=.configs/pylintrc --jobs=0 --exit-zero $(< $TRAVIS_HOME/before_files.txt) > $TRAVIS_HOME/pylint_before_output.txt
```
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply .travis.yml]: Great

File: .travis.yml [hunk 3]
```diff
@@ -1,24 +1,70 @@
...
+              - >-
+                grep -F "/10, -" $TRAVIS_HOME/pylint_before_output.txt ||
+                grep -F "/10, +0.00" $TRAVIS_HOME/pylint_before_output.txt ||
+                (echo "pylint score decreased, please try again after fixing some lint issues." && cat $TRAVIS_HOME/pylint_after_output.txt && false)
```
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply .travis.yml]: Ok, let's see how it turns out then

**Pull request merged**